// @serverless-stack/resources/dist/util/customResource.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCustomResource = void 0;
const cdk = __importStar(require("@aws-cdk/core"));
const lambda = __importStar(require("@aws-cdk/aws-lambda"));
function createCustomResource(scope, id, props) {
    // Create handler
    const handler = new lambda.Function(scope, `${id}Provider`, {
        code: lambda.Code.fromAsset(props.srcPath),
        handler: props.handler,
        description: props.description,
        runtime: lambda.Runtime.NODEJS_12_X,
        timeout: cdk.Duration.minutes(15),
        memorySize: 1024,
        initialPolicy: props.policyStatements,
    });
    // Create custom resource
    const resource = new cdk.CustomResource(scope, id, {
        resourceType: props.resourceType,
        serviceToken: handler.functionArn,
        properties: props.properties,
    });
    return { handler, resource };
}
exports.createCustomResource = createCustomResource;
